flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceSplitEnumerator.java

```diff
@@ -218,7 +218,8 @@ public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
             }
 
             if (subtaskSourceIndex < currentSourceIndex) {
-                subtaskSourceIndex++;
+                // find initial or next index for the reader
+                subtaskSourceIndex = switchedSources.getNextSubtaskSourceIndex(subtaskSourceIndex);
```

Comment:
IMO, the `getNextSubtaskSourceIndex` method name isn't matched for `SwitchedSources`. There is no any subtask concept in `SwitchedSources`.